### PR TITLE
Implemented Controls.js superclass with pointer events and multi-touch drag tracking

### LINKED ISSUE #20191: Clicking on the Ammo sample in the r120 does not launch the iron ball
##### Description of the problem

In the previous Ammo sample, clicking with the mouse launch an iron ball.
However, in the `r120` [sample](https://github.com/mrdoob/three.js/blob/r120/examples/physics_ammo_break.html), the iron ball does not appear when the mouse is clicked.
I tried the combination of the r120 sample and the previous version of [OrbitControls.js](https://github.com/mrdoob/three.js/blob/r119/examples/jsm/controls/OrbitControls.js) and the iron ball does appear, so I think this control is probably related.

[r120 Ammo Convex object breaking example result:](https://rawcdn.githack.com/mrdoob/three.js/r120/examples/physics_ammo_break.html)
image

[r119 Ammo Convex object breaking example result:](https://rawcdn.githack.com/mrdoob/three.js/r119/examples/physics_ammo_break.html)
image

##### Three.js version

- [x] r120

##### Browser

- [x] Chrome
- [x] Firefox

##### OS

- [x] Windows

##### Hardware Requirements (graphics card, VR Device, ...)

MacBook Air (BootCamp) Intel(R) UHD Graphics 617 Direct3D11 vs_5_0 ps_5_0

### PR DESCRIPTION
**Pull request by arodic**
Implemented generic Controls class.

Adds/removes event listeners during lifecycle and on `enabled` property change.

Forwards pointer and keyboard events:
`"contextmenu"`, `"wheel"`, `"pointerdown"`, `"pointermove"`, `"pointerup"`, `"keydown"`, `"keyup"`

Evokes methods with tracked pointer and keyCode data:
`onTrackedPointerDown`, `onTrackedPointerMove`, `onTrackedPointerUp`, `onTrackedKeyDown`, `onTrackedKeyUp`, `onTrackedKeyChange`.

Refactored OrbitControls, MapControls, DragControls, TrackballControls to inherit Controls superclass.
    
Improved DragControls to work with multiouch - removed raycast while dragging.

### CODE REVIEW

**Comment by arodic** [comment]
To make [pointerEvents work with multi-touch efficiently](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events/Using_Pointer_Events), developers should manage and track multiple pointers in arrays for example. Since pointer ID's are unique and consistent for each interaction it is easy to do so.

For this, I created a simple `ControlPointer` class but I'm not sure what is the best place to put it for easier reuse. I can see this being useful beyond controls.

Notice that this PR removes 200 lines of code!

**Comment by Mugen87** [comment]
@arodic I've converted your PR to a "Draft PR" (which seems ideal for this use case).

**Comment by DefinitelyMaybe** [comment]
if you aim at the `examples/js` folders scripts you'll have a much more luck getting this PR out of draft and merged into dev

**Comment by arodic** [comment]
@DefinitelyMaybe thanks for the heads up!

**Comment by arodic** [comment]
I'm having too much fun with this. Here is another suggested enhancement for Controls.js io-gui/three.js/pull#1

It makes extending Controls.js easier and leaves event handling untouched and more flexible. This new version only takes care of event listener setup/teardown, reacts to disabled property change and it captures and tracks active pointers.

Let me know what you think,  I might have some more time to work on this in a couple of weeks.

**Comment by mrdoob** [comment]
I'm sorry Aki, but I don't think these files should have dependencies.

**Comment by arodic** [comment]
@DefinitelyMaybe I moved the code to examples/js but since these versions require Controls.js as a dependency I also had to [hack a line in modularize tool](https://github.com/mrdoob/three.js/pull/20219/files#diff-6e2dbedb9f77ae9102b96614de9b0102R254).

> I'm sorry Aki, but I don't think these files should have dependencies. Remember to keep things simple.

That's a bummer @mrdoob :( But I still hope you will change your mind once examples are fully transitioned to ES6 modules, because simplicity is EXACTLY what I am after. In my opinion, one of the best ways to "keep things simple" is to encapsulate complexity and articulate it with a simple API through inheritance. Another simplifying effect of inheritance is the fact that you dont have to fix the same problem multiple times. 

To illustrate how this change adds simplicity, I also converted `DragControls` to use `Controls` superclass. Not only this reduced the amount of code in `DragControls` by 77 lines, it also automatically added support for muti-touch drag and it fixed some performance issues while making the code a bit simpler.

The size of `OrbitControls` is reduced by 339 lines and its quite a bit simpler.

**Comment by mrdoob** [comment]
We are currently trying to adopt Pointer Events in all controls. Once we are done with that, then PRs like this one could be considered. I prefer to go one step at a time, rather than many steps at a time. I hope you understand.

File: examples/jsm/controls/Controls.js [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+		this.ctrlKey = pointerEvent.ctrlKey;
+		this.metaKey = pointerEvent.metaKey;
+		this.shiftKey = pointerEvent.shiftKey;
+
+		this.previous.copy( this.client );
+		this.client.set( pointerEvent.clientX, pointerEvent.clientY );
+		this.page.set( pointerEvent.pageX, pointerEvent.pageY );
+		// Calculate movement because Safari movement is broken
+		this.movement.copy( this.client ).sub( this.previous );
```
Comment (arodic) [inline examples/jsm/controls/Controls.js:334]: Controls.js now includes manual movement vector calculation since movementX/Y is broken in Safari.

**Comment by arodic** [comment]
Closing this PR since it's not getting any traction. Might give it another shot in the future.

**Pull request closed without merging**